libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,284 @@
+							var.name = m_nameDispenser.newName(var.name);
+							mstores.emplace_back(getMemoryLocation(slot), var.name);
+						}
+					result.emplace_back(std::move(_statement));
+					appendMemoryStores(result, loc, mstores);
+					return {std::move(result)};
+				}
+				else
+				{
+					ASTModifier::visit(_statement);
+					return {};
+				}
+			}
+		);
+	}
+	virtual void visit(Expression& _expression)
+	{
+		if (
+			auto identifier = std::get_if<Identifier>(&_expression);
+			identifier && m_currentFunctionMemorySlots && m_currentFunctionMemorySlots->count(identifier->name)
+		)
+		{
+			auto loc = identifier->location;
+			_expression = FunctionCall {
+				loc,
+				Identifier{loc, "mload"_yulstring}, {
+					Literal {
+						loc,
+						LiteralKind::Number,
+						getMemoryLocation(m_currentFunctionMemorySlots->at(identifier->name)),
+						{}
+					}
+				}
+			};
+		}
+		else
+			ASTModifier::visit(_expression);
+	}
+private:
+	std::map<YulString, std::map<YulString, uint64_t>> const& m_memorySlots;
+	u256 m_reservedMemory;
+	NameDispenser& m_nameDispenser;
+	std::map<YulString, uint64_t> const* m_currentFunctionMemorySlots = nullptr;
+};
+
+}
+
+void MemoryEscalator::run(OptimiserStepContext& _context, Object& _object, bool _optimizeStackAllocation)
+{
+	auto functionStackErrorInfo = CompilabilityChecker::run(_context.dialect, _object, _optimizeStackAllocation);
+	if (functionStackErrorInfo.empty())
+		return;
+
+	yulAssert(dynamic_cast<EVMDialect const*>(&_context.dialect), "MemoryEscalator can only be run on EVMDialect objects.");
+	// TODO: make this more robust with value tracking and stepping through the outermost block
+	FunctionCall* memoryInit = findMemoryInit(_object);
+	if (!memoryInit)
+		return;
+	u256 reservedMemory = 0;
+	if (auto reservedMemoryLiteral = std::get_if<Literal>(&memoryInit->arguments.back()))
+		reservedMemory = valueOfLiteral(*reservedMemoryLiteral);
+	else
+		return;
+
+	auto callGraph = CallGraphGenerator::callGraph(*_object.code).functionCalls;
+
+	std::set<YulString> leaves;
+
+	// TODO: I'm assuming it doesn't contain directed cycles so far. Need to check and exit otherwise.
+	std::map<YulString, set<YulString>> backEdges;
+	util::BreadthFirstSearch<YulString>{{{}}}.run([&](YulString const& _node, auto&& _addChild) {
+		if (callGraph.count(_node))
+		{
+			auto const& calledFunctions = callGraph.at(_node);
+			if (calledFunctions.empty())
+				leaves.emplace(_node);
+			else
+				for (auto const& calledFunction: callGraph.at(_node))
+				{
+					backEdges[calledFunction].insert(_node);
+					_addChild(calledFunction);
+				}
+		}
+		else
+			leaves.emplace(_node);
+	});
+
+	std::map<YulString, std::map<YulString, uint64_t>> assignedMemorySlots;
+	uint64_t requiredSlots = 0;
+	{
+		std::set<YulString> visited;
+		std::map<YulString, uint64_t> nextAvailableSlots;
+		std::list<YulString> V{leaves.begin(), leaves.end()};
+		while (!V.empty())
```

Comment:
I'm wondering if it might be possible to write a recursive algorithm for this instead.

Reply:
Would that be preferable? It should definitely be possible, but I actually tend to favor loops over recursion in order not to put too much strain on the call stack...

Reply:
Variables are identified by `pair<YulString, YulString>` - function name, variable name.

This ignores back edges.
```
using SlotAllocations = map<pair<YulString, YulString>, size_t>;
SlotAllocations allocateMemorySlots(FuntionDefinition const& _f, SlotAllocations _slotAllocations)
{
  auto vars = variablesInNeedOfEscalation(f);
  auto currentSlotAllocations = addVarsToAllocations(_slotAllocations, vars);
  auto newSlotAllocations = currentSlotAllocations;
  for (child: callees(f))
    newSlotAllocations += allocateMemorySlots(child, currentSlotAllocations);
  return newSlotAllocations;
}
```

Reply:
So allocating *forward* starting from the root rather than backwards starting from leaves? I'll have to think it through, but it might work. I think the main point in doing it the other way round was that we could save memory by moving the free memory init down the call graph later on (for that it's better if the leaves have lower slots and near-to-root gets higher slots) - but if we don't do that anyways, this might be simpler...